src/main/java/net/minecraftforge/registries/RegistryBuilder.java

```diff
@@ -230,6 +244,57 @@ public RegistryBuilder<T> hasTags()
         return this;
     }
 
+    /**
+     * <p>Register this registry as an unsynced datapack registry, which will cause data to be loaded from
+     * a datapack folder based on the registry's name. The mod that registers this registry does not need to exist
+     * on the client to connect to servers with the mod/registry.</p>
+     * <p>The registry's name must be of the form {@code modid:modid/folder}.
+     * Data JSONs will be loaded from {@code data/<datapack_namespace>/modid/folder/}, where modid is the mod that registered this registry.</p>
+     * 
+     * @param codec the codec to be used for loading data from datapacks on servers
+     * @return this builder
+     * 
+     * @see {@link #dataPackRegistry(Codec, Codec)} for registering a synced datapack registry
+     */
+    public RegistryBuilder<T> dataPackRegistry(Codec<T> codec)
+    {
+        return this.dataPackRegistry(codec, null);
+    }
+
+    /**
+     * <p>Register this registry as a datapack registry, which will cause data to be loaded from
+     * a datapack folder based on the registry's name.</p>
+     * <p>The registry's name must be of the form {@code modid:modid/folder}.
+     * Data JSONs will be loaded from {@code data/<datapack_namespace>/modid/folder/}, where modid is the mod that registered this registry.</p>
+     * 
+     * @param codec the codec to be used for loading data from datapacks on servers
+     * @param networkCodec the codec to be used for syncing loaded data to clients.<br>
+     * If networkCodec is null, data will not be synced, and clients without the mod that registered this registry can
+     * connect to servers with the mod.<br>
+     * If networkCodec is not null, then data will be synced (accessible via {@link ClientPacketListener#registryAccess()}),
+     * and the mod must be present on a client to connect to servers with the mod.
+     * @return this builder
+     * 
+     * @see {@link #dataPackRegistry(Codec)} for registering an unsynced datapack registry
+     */
+    public RegistryBuilder<T> dataPackRegistry(Codec<T> codec, @Nullable Codec<T> networkCodec)
+    {
+        this.hasWrapper(); // A wrapper is required for data pack registries.
+        this.disableSync(); // Datapack registries are synced using a different system than static registries.
+        // Supplier averts having to set the registry name before calling this.
+        this.dataPackRegistryData = Suppliers.memoize(() -> {
+            // Validate registry key.
+            if (this.registryName == null)
+                throw new IllegalStateException("Registry builder cannot build a datapack registry: registry name not set");
+            if (!DataPackRegistriesHooks.isNamespacedRegistryFolder(this.registryName))
+                throw new IllegalStateException(String.format(Locale.ENGLISH, "Invalid datapack registry name %s: Registry path must be prefixed with '%s'", registryName, registryName.getNamespace() + '/'));
```

Comment:
Should we document these being thrown in the javadoc (noting also that these are thrown when the registry is actually built, not thrown at this method's invocation)?

Reply:
We can add a throws doc in getDataPackRegistryData, since that actually will throw on invocation if registry name hasn't been set yet (my understanding is that all registries need to be named eventually anyway so saying "make sure to call setRegistryName eventually" here is probably redundant).